File: tests/python/test_jma.py
"""
Python binding tests for JMA indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestJma:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_jma_partial_params(self, test_data):
        """Test JMA with partial parameters - mirrors check_jma_partial_params"""
        close = test_data['close']
        
        
        result = ta_indicators.jma(close, 7)
        assert len(result) == len(close)
    
    def test_jma_accuracy(self, test_data):
        """Test JMA matches expected values from Rust tests - mirrors check_jma_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['jma']
        
        
        result = ta_indicators.jma(
            close,
            period=expected['default_params']['period'],
            phase=expected['default_params']['phase'],
            power=expected['default_params']['power']
        )
        
        assert len(result) == len(close)
        
        
        
        assert_close(
            result[-5:], 
            expected['last_5_values'],
            rtol=0,
            atol=1e-6,
            msg="JMA last 5 values mismatch"
        )
        
        
        compare_with_rust('jma', result, 'close', expected['default_params'])
    
    def test_jma_default_candles(self, test_data):
        """Test JMA with default parameters - mirrors check_jma_default_candles"""
        close = test_data['close']
        
        
        result = ta_indicators.jma(close, 7, 50.0, 2)
        assert len(result) == len(close)
        
        
        
    
    def test_jma_zero_period(self):
        """Test JMA fails with zero period - mirrors check_jma_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError):
            ta_indicators.jma(input_data, period=0)
    
    def test_jma_period_exceeds_length(self):
        """Test JMA fails when period exceeds data length - mirrors check_jma_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError):
            ta_indicators.jma(data_small, period=10)
    
    def test_jma_very_small_dataset(self):
        """Test JMA with very small dataset - mirrors check_jma_very_small_dataset"""
        data_single = np.array([42.0])
        
        with pytest.raises(ValueError):
            ta_indicators.jma(data_single, period=7)
    
    def test_jma_empty_input(self):
        """Test JMA with empty input"""
        data_empty = np.array([])
        
        with pytest.raises(ValueError):
            ta_indicators.jma(data_empty, period=7)
    
    def test_jma_all_nan(self):
        """Test JMA with all NaN input"""
        data = np.array([np.nan, np.nan, np.nan, np.nan, np.nan])
        
        with pytest.raises(ValueError):
            ta_indicators.jma(data, period=3)
    
    def test_jma_invalid_phase(self):
        """Test JMA with invalid phase values"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
        
        
        with pytest.raises(ValueError):
            ta_indicators.jma(data, period=3, phase=float('nan'))
        
        
        with pytest.raises(ValueError):
            ta_indicators.jma(data, period=3, phase=float('inf'))
    
    def test_jma_batch_with_nan_data(self):
        """Test JMA batch with NaN prefix data"""
        
        data = np.array([np.nan, np.nan] + list(range(1, 21)))
        
        result = ta_indicators.jma_batch(
            data,
            period_range=(3, 5, 2),  
            phase_range=(50.0, 50.0, 0.0),
            power_range=(2, 2, 0)
        )
        
        values = result['values']
        assert values.shape == (2, len(data))
        
        
        for row in range(2):
            assert np.isnan(values[row, 0]), f"Row {row} index 0 should be NaN"
            assert np.isnan(values[row, 1]), f"Row {row} index 1 should be NaN"
            
            assert np.isfinite(values[row, 2]), f"Row {row} index 2 should be finite"
    
    def test_jma_nan_handling(self, test_data):
        """Test JMA handling of NaN values - mirrors check_jma_nan_handling"""
        
        data_with_nan = np.concatenate([
            np.array([np.nan, np.nan, np.nan]),
            test_data['close'][:50]
        ])
        period = 7
        
        result = ta_indicators.jma(data_with_nan, period)
        
        assert len(result) == len(data_with_nan)
        
        
        assert np.all(np.isnan(result[:3])), "Expected NaN before first valid data"
        
        
        assert np.isfinite(result[3]), "Expected finite value at first valid"
        
        
        finite_count = np.sum(np.isfinite(result[3:]))
        assert finite_count == len(result[3:]), "All values after first valid should be finite"
    
    def test_jma_streaming(self, test_data):
        """Test JMA streaming vs batch calculation - mirrors check_jma_streaming"""
        close = test_data['close'][:100]  
        period = 7
        phase = 50.0
        power = 2
        
        
        batch_result = ta_indicators.jma(close, period, phase, power)
        
        
        stream = ta_indicators.JmaStream(period, phase, power)
        stream_results = []
        
        for val in close:
            result = stream.update(val)
            stream_results.append(result if result is not None else np.nan)
        
        stream_results = np.array(stream_results)
        
        
        assert_close(
            stream_results, 
            batch_result,
            rtol=1e-10,
            msg="JMA streaming vs batch mismatch"
        )
    
    def test_jma_batch(self, test_data):
        """Test JMA batch computation with proper warmup verification."""
        close = test_data['close']
        
        
        period_range = (5, 9, 2)  
        phase_range = (40.0, 60.0, 10.0)  
        power_range = (1, 3, 1)  
        
        result = ta_indicators.jma_batch(close, period_range, phase_range, power_range)
        
        assert isinstance(result, dict)
        assert 'values' in result
        assert 'periods' in result
        assert 'phases' in result
        assert 'powers' in result
        
        values = result['values']
        periods = result['periods']
        phases = result['phases']
        powers = result['powers']
        
        
        assert values.shape == (27, len(close))
        assert len(periods) == 27
        assert len(phases) == 27
        assert len(powers) == 27
        
        
        individual_result = ta_indicators.jma(close, 5, 40.0, 1)
        np.testing.assert_allclose(
            values[0], 
            individual_result, 
            rtol=1e-9,
            err_msg="Batch first row mismatch"
        )
        
        
        
        for i in range(27):
            row = values[i]
            
            assert np.all(np.isfinite(row)), f"Row {i} has unexpected NaN values"
    
    def test_jma_different_params(self, test_data):
        """Test JMA with different parameter values."""
        close = test_data['close']
        
        
        param_sets = [
            (5, 0.0, 1),     
            (7, 50.0, 2),    
            (10, 100.0, 3),  
            (14, -100.0, 2), 
        ]
        
        for period, phase, power in param_sets:
            result = ta_indicators.jma(close, period, phase, power)
            assert len(result) == len(close)
            
            
            finite_count = np.sum(np.isfinite(result))
            assert finite_count > len(close) - period * 2, \
                f"Too many non-finite values for params ({period}, {phase}, {power})"
    
    def test_jma_batch_performance(self, test_data):
        """Test that batch computation is more efficient than multiple single computations."""
        close = test_data['close'][:1000]  
        
        
        import time
        
        start_batch = time.time()
        batch_result = ta_indicators.jma_batch(
            close, 
            (5, 7, 2),      
            (40.0, 50.0, 10.0),  
            (1, 2, 1)       
        )
        batch_time = time.time() - start_batch
        
        start_single = time.time()
        single_results = []
        for period in [5, 7]:
            for phase in [40.0, 50.0]:
                for power in [1, 2]:
                    single_results.append(ta_indicators.jma(close, period, phase, power))
        single_time = time.time() - start_single
        
        
        print(f"Batch time: {batch_time:.4f}s, Single time: {single_time:.4f}s")
        
        
        values = batch_result['values']
        for i, single in enumerate(single_results):
            np.testing.assert_allclose(values[i], single, rtol=1e-9)
    
    def test_jma_phase_range(self):
        """Test JMA with full phase range"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0])
        
        
        for phase in [-100.0, -50.0, 0.0, 50.0, 100.0]:
            result = ta_indicators.jma(data, period=3, phase=phase, power=2)
            assert len(result) == len(data)
            assert np.sum(np.isfinite(result)) > 0
    
    def test_jma_power_values(self):
        """Test JMA with different power values"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0])
        
        
        for power in [1, 2, 3, 5, 10]:
            result = ta_indicators.jma(data, period=3, phase=50.0, power=power)
            assert len(result) == len(data)
            assert np.sum(np.isfinite(result)) > 0
    
    def test_jma_single_param_batch(self, test_data):
        """Test JMA batch with single parameter variation"""
        close = test_data['close'][:100]
        
        
        result = ta_indicators.jma_batch(
            close,
            (5, 9, 2),         
            (50.0, 50.0, 0.0), 
            (2, 2, 0)          
        )
        
        values = result['values']
        periods = result['periods']
        phases = result['phases']
        powers = result['powers']
        
        
        assert values.shape == (3, len(close))
        assert list(periods) == [5, 7, 9]
        assert all(p == 50.0 for p in phases)
        assert all(p == 2 for p in powers)
    
    def test_jma_batch_metadata(self, test_data):
        """Test batch metadata function returns correct parameter combinations"""
        period_range = (3, 5, 2)     
        phase_range = (40.0, 50.0, 10.0)  
        power_range = (1, 2, 1)      
        
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0])
        result = ta_indicators.jma_batch(data, period_range, phase_range, power_range)
        
        periods = result['periods']
        phases = result['phases']
        powers = result['powers']
        
        
        assert len(periods) == 8
        assert len(phases) == 8
        assert len(powers) == 8
        
        
        expected_combinations = [
            (3, 40.0, 1),
            (3, 40.0, 2),
            (3, 50.0, 1),
            (3, 50.0, 2),
        ]
        
        for i, (exp_period, exp_phase, exp_power) in enumerate(expected_combinations):
            assert periods[i] == exp_period
            assert abs(phases[i] - exp_phase) < 1e-9
            assert powers[i] == exp_power
    
    def test_jma_warmup_behavior(self, test_data):
        """Test JMA warmup behavior - outputs at first_valid with NaN prefix"""
        
        data = np.array([np.nan, np.nan, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0])
        period = 3
        
        result = ta_indicators.jma(data, period=period, phase=50.0, power=2)
        
        
        assert np.isnan(result[0]), "Index 0 should be NaN"
        assert np.isnan(result[1]), "Index 1 should be NaN"
        
        
        assert np.isfinite(result[2]), "Index 2 (first valid) should have a value"
        
        
        for i in range(3, len(result)):
            assert np.isfinite(result[i]), f"Expected finite value at index {i}"
        
        
        clean_data = test_data['close'][:20]
        clean_result = ta_indicators.jma(clean_data, period=7)
        
        
        assert np.isfinite(clean_result[0]), "JMA should output at first valid (index 0 for clean data)"


    def test_jma_consistency_single_vs_batch(self, test_data):
        """Test that single and batch calculations are consistent"""
        close = test_data['close'][:100]
        
        
        single_result = ta_indicators.jma(close, 7, 50.0, 2)
        
        
        batch_result = ta_indicators.jma_batch(
            close,
            period_range=(7, 7, 0),
            phase_range=(50.0, 50.0, 0.0),
            power_range=(2, 2, 0)
        )
        
        
        batch_single = batch_result['values'][0]
        
        
        np.testing.assert_allclose(
            single_result,
            batch_single,
            rtol=1e-12,
            err_msg="Single vs batch calculation mismatch"
        )


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: tests/wasm/test_jma.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('JMA partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.jma_js(close, 7, 50.0, 2);
    assert.strictEqual(result.length, close.length);
});

test('JMA accuracy', async () => {

    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.jma;

    const result = wasm.jma_js(
        close,
        expected.defaultParams.period,
        expected.defaultParams.phase,
        expected.defaultParams.power
    );

    assert.strictEqual(result.length, close.length);


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-6,
        "JMA last 5 values mismatch"
    );


    await compareWithRust('jma', result, 'close', expected.defaultParams);
});

test('JMA default candles', async () => {

    const close = new Float64Array(testData.close);

    const result = wasm.jma_js(close, 7, 50.0, 2);
    assert.strictEqual(result.length, close.length);


    await compareWithRust('jma', result, 'close', { period: 7, phase: 50.0, power: 2 });
});

test('JMA zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.jma_js(inputData, 0, 50.0, 2);
    });
});

test('JMA period exceeds length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.jma_js(dataSmall, 10, 50.0, 2);
    });
});

test('JMA very small dataset', () => {

    const dataSingle = new Float64Array([42.0]);

    assert.throws(() => {
        wasm.jma_js(dataSingle, 7, 50.0, 2);
    });
});

test('JMA empty input', () => {

    const dataEmpty = new Float64Array([]);

    assert.throws(() => {
        wasm.jma_js(dataEmpty, 7, 50.0, 2);
    });
});

test('JMA all NaN', () => {

    const data = new Float64Array([NaN, NaN, NaN, NaN, NaN]);

    assert.throws(() => {
        wasm.jma_js(data, 3, 50.0, 2);
    });
});

test('JMA invalid phase', () => {

    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);


    assert.throws(() => {
        wasm.jma_js(data, 3, NaN, 2);
    });


    assert.throws(() => {
        wasm.jma_js(data, 3, Infinity, 2);
    });
});

test('JMA reinput', () => {

    const close = new Float64Array(testData.close);


    const firstResult = wasm.jma_js(close, 7, 50.0, 2);


    const secondResult = wasm.jma_js(firstResult, 7, 50.0, 2);

    assert.strictEqual(secondResult.length, firstResult.length);
});

test('JMA NaN handling', () => {

    const dataWithNan = new Float64Array([NaN, NaN, NaN, ...testData.close.slice(0, 50)]);
    const period = 7;

    const result = wasm.jma_js(dataWithNan, period, 50.0, 2);

    assert.strictEqual(result.length, dataWithNan.length);


    assertAllNaN(result.slice(0, 3), "Expected NaN before first valid data");


    assert(isFinite(result[3]), "Expected finite value at first valid");


    for (let i = 3; i < result.length; i++) {
        assert(isFinite(result[i]), `Expected finite value at index ${i}`);
    }
});

test('JMA NaN handling (original)', () => {

    const close = new Float64Array(testData.close);
    const period = 7;

    const result = wasm.jma_js(close, period, 50.0, 2);

    assert.strictEqual(result.length, close.length);


    let finiteCount = 0;
    for (let val of result) {
        if (isFinite(val)) finiteCount++;
    }
    assert(finiteCount > close.length - period * 2, "Too many non-finite values");
});

test('JMA batch', () => {

    const close = new Float64Array(testData.close);


    const period_start = 5, period_end = 9, period_step = 2;
    const phase_start = 40.0, phase_end = 60.0, phase_step = 10.0;
    const power_start = 1, power_end = 3, power_step = 1;

    const batch_result = wasm.jma_batch_js(
        close,
        period_start, period_end, period_step,
        phase_start, phase_end, phase_step,
        power_start, power_end, power_step
    );
    const metadata = wasm.jma_batch_metadata_js(
        period_start, period_end, period_step,
        phase_start, phase_end, phase_step,
        power_start, power_end, power_step
    );


    assert.strictEqual(metadata.length, 27 * 3);


    assert.strictEqual(batch_result.length, 27 * close.length);


    const individual_result = wasm.jma_js(close, 5, 40.0, 1);


    const row = batch_result.slice(0, close.length);

    assertArrayClose(row, individual_result, 1e-9, 'First combination');
});

test('JMA different params', () => {

    const close = new Float64Array(testData.close);


    const paramSets = [
        [5, 0.0, 1],
        [7, 50.0, 2],
        [10, 100.0, 3],
        [14, -100.0, 2],
    ];

    for (const [period, phase, power] of paramSets) {
        const result = wasm.jma_js(close, period, phase, power);
        assert.strictEqual(result.length, close.length);


        let finiteCount = 0;
        for (let val of result) {
            if (isFinite(val)) finiteCount++;
        }
        assert(finiteCount > close.length - period * 2,
            `Too many non-finite values for params (${period}, ${phase}, ${power})`);
    }
});

test('JMA batch performance', () => {

    const close = new Float64Array(testData.close.slice(0, 1000));


    const startBatch = performance.now();
    const batchResult = wasm.jma_batch_js(
        close,
        5, 7, 2,
        40.0, 50.0, 10.0,
        1, 2, 1
    );
    const batchTime = performance.now() - startBatch;

    const startSingle = performance.now();
    const singleResults = [];
    for (let period = 5; period <= 7; period += 2) {
        for (let phase = 40.0; phase <= 50.0; phase += 10.0) {
            for (let power = 1; power <= 2; power += 1) {
                singleResults.push(...wasm.jma_js(close, period, phase, power));
            }
        }
    }
    const singleTime = performance.now() - startSingle;


    console.log(`Batch time: ${batchTime.toFixed(2)}ms, Single time: ${singleTime.toFixed(2)}ms`);


    assertArrayClose(batchResult, singleResults, 1e-9, 'Batch vs single results');
});

test('JMA phase range', () => {

    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0]);


    for (const phase of [-100.0, -50.0, 0.0, 50.0, 100.0]) {
        const result = wasm.jma_js(data, 3, phase, 2);
        assert.strictEqual(result.length, data.length);

        let finiteCount = 0;
        for (let val of result) {
            if (isFinite(val)) finiteCount++;
        }
        assert(finiteCount > 0, `No finite values for phase ${phase}`);
    }
});

test('JMA power values', () => {

    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0]);


    for (const power of [1, 2, 3, 5, 10]) {
        const result = wasm.jma_js(data, 3, 50.0, power);
        assert.strictEqual(result.length, data.length);

        let finiteCount = 0;
        for (let val of result) {
            if (isFinite(val)) finiteCount++;
        }
        assert(finiteCount > 0, `No finite values for power ${power}`);
    }
});

test('JMA single param batch', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const batch_result = wasm.jma_batch_js(
        close,
        5, 9, 2,
        50.0, 50.0, 0.0,
        2, 2, 0
    );


    assert.strictEqual(batch_result.length, 3 * close.length);


    for (let i = 0; i < 3; i++) {
        const period = 5 + i * 2;
        const start = i * close.length;
        const row = batch_result.slice(start, start + close.length);

        const individual = wasm.jma_js(close, period, 50.0, 2);
        assertArrayClose(row, individual, 1e-9, `Period ${period}`);
    }
});

test('JMA batch metadata', () => {

    const metadata = wasm.jma_batch_metadata_js(
        5, 7, 2,
        40.0, 50.0, 10.0,
        1, 2, 1
    );



    assert.strictEqual(metadata.length, 24);


    const expectedFirstCombinations = [
        5, 40.0, 1,
        5, 40.0, 2,
        5, 50.0, 1,
        5, 50.0, 2,
    ];

    for (let i = 0; i < expectedFirstCombinations.length; i++) {
        assertClose(
            metadata[i],
            expectedFirstCombinations[i],
            1e-9,
            `Metadata value at index ${i}`
        );
    }
});

test('JMA warmup behavior', () => {


    const data = new Float64Array([NaN, NaN, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0]);
    const period = 3;

    const result = wasm.jma_js(data, period, 50.0, 2);


    assert(isNaN(result[0]), "Index 0 should be NaN");
    assert(isNaN(result[1]), "Index 1 should be NaN");


    assert(isFinite(result[2]), "Index 2 (first valid) should have a value");


    for (let i = 3; i < result.length; i++) {
        assert(isFinite(result[i]), `Expected finite value at index ${i}`);
    }


    const cleanData = new Float64Array(testData.close.slice(0, 20));
    const cleanResult = wasm.jma_js(cleanData, 7, 50.0, 2);


    assert(isFinite(cleanResult[0]), "JMA should output at first valid (index 0 for clean data)");
});

test('JMA consistency across calls', () => {

    const close = new Float64Array(testData.close.slice(0, 100));

    const result1 = wasm.jma_js(close, 7, 50.0, 2);
    const result2 = wasm.jma_js(close, 7, 50.0, 2);

    assertArrayClose(result1, result2, 1e-15, "JMA results not consistent");
});

test('JMA parameter boundaries', () => {

    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);


    const result1 = wasm.jma_js(data, 1, 50.0, 2);
    assert.strictEqual(result1.length, data.length);


    const smallData = new Float64Array([1, 2, 3, 4, 5]);
    const result2 = wasm.jma_js(smallData, 4, 50.0, 2);
    assert.strictEqual(result2.length, smallData.length);
});


test('JMA zero-copy API', () => {
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    const period = 3;
    const phase = 50.0;
    const power = 2;


    const ptr = wasm.jma_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');


    const memory = wasm.__wasm ? wasm.__wasm.memory : (wasm.__wbindgen_memory ? wasm.__wbindgen_memory() : wasm.memory);
    const memView = new Float64Array(
        memory.buffer,
        ptr,
        data.length
    );


    memView.set(data);


    try {
        wasm.jma_into(ptr, ptr, data.length, period, phase, power);


        const regularResult = wasm.jma_js(data, period, phase, power);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue;
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {

        wasm.jma_free(ptr, data.length);
    }
});

test('JMA zero-copy with NaN data', () => {
    const data = new Float64Array([NaN, NaN, 1, 2, 3, 4, 5, 6, 7, 8]);
    const period = 3;
    const phase = 50.0;
    const power = 2;

    const ptr = wasm.jma_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');

    try {
        const memory = wasm.__wasm ? wasm.__wasm.memory : (wasm.__wbindgen_memory ? wasm.__wbindgen_memory() : wasm.memory);
        const memView = new Float64Array(memory.buffer, ptr, data.length);
        memView.set(data);

        wasm.jma_into(ptr, ptr, data.length, period, phase, power);


        const memory2 = wasm.__wasm ? wasm.__wasm.memory : (wasm.__wbindgen_memory ? wasm.__wbindgen_memory() : wasm.memory);
        const memView2 = new Float64Array(memory2.buffer, ptr, data.length);


        assert(isNaN(memView2[0]), "Expected NaN at index 0");
        assert(isNaN(memView2[1]), "Expected NaN at index 1");


        assert(isFinite(memView2[2]), "Expected finite value at first valid (index 2)");
    } finally {
        wasm.jma_free(ptr, data.length);
    }
});

test('JMA zero-copy error handling', () => {

    assert.throws(() => {
        wasm.jma_into(0, 0, 10, 3, 50.0, 2);
    }, /null pointer/i);


    const ptr = wasm.jma_alloc(10);
    try {

        assert.throws(() => {
            wasm.jma_into(ptr, ptr, 10, 0, 50.0, 2);
        }, /Invalid period/);


        assert.throws(() => {
            wasm.jma_into(ptr, ptr, 10, 3, NaN, 2);
        }, /Invalid phase/);
    } finally {
        wasm.jma_free(ptr, 10);
    }
});

test('JMA zero-copy memory management', () => {

    const sizes = [100, 1000, 10000];

    for (const size of sizes) {
        const ptr = wasm.jma_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);


        const memory = wasm.__wasm ? wasm.__wasm.memory : (wasm.__wbindgen_memory ? wasm.__wbindgen_memory() : wasm.memory);
        const memView = new Float64Array(memory.buffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }


        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }


        wasm.jma_free(ptr, size);
    }
});


test('JMA batch - new ergonomic API with single parameter', () => {
    const close = new Float64Array(testData.close.slice(0, 100));

    const result = wasm.jma_batch(close, {
        period_range: [7, 7, 0],
        phase_range: [50.0, 50.0, 0],
        power_range: [2, 2, 0]
    });


    assert(result.values, 'Should have values array');
    assert(result.combos, 'Should have combos array');
    assert(typeof result.rows === 'number', 'Should have rows count');
    assert(typeof result.cols === 'number', 'Should have cols count');


    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, close.length);
    assert.strictEqual(result.combos.length, 1);
    assert.strictEqual(result.values.length, close.length);


    const combo = result.combos[0];
    assert.strictEqual(combo.period, 7);
    assert.strictEqual(combo.phase, 50.0);
    assert.strictEqual(combo.power, 2);


    const singleResult = wasm.jma_js(close, 7, 50.0, 2);
    assertArrayClose(result.values, singleResult, 1e-10, 'Batch vs single mismatch');
});

test('JMA batch - new API with multiple parameters', () => {
    const close = new Float64Array(testData.close.slice(0, 50));

    const result = wasm.jma_batch(close, {
        period_range: [5, 7, 2],
        phase_range: [40.0, 50.0, 10.0],
        power_range: [2, 2, 0]
    });


    assert.strictEqual(result.rows, 4);
    assert.strictEqual(result.cols, 50);
    assert.strictEqual(result.combos.length, 4);
    assert.strictEqual(result.values.length, 200);


    const expectedCombos = [
        { period: 5, phase: 40.0, power: 2 },
        { period: 5, phase: 50.0, power: 2 },
        { period: 7, phase: 40.0, power: 2 },
        { period: 7, phase: 50.0, power: 2 }
    ];

    for (let i = 0; i < expectedCombos.length; i++) {
        assert.strictEqual(result.combos[i].period, expectedCombos[i].period);
        assert.strictEqual(result.combos[i].phase, expectedCombos[i].phase);
        assert.strictEqual(result.combos[i].power, expectedCombos[i].power);
    }


    const firstRow = result.values.slice(0, result.cols);
    const singleResult = wasm.jma_js(close, 5, 40.0, 2);
    assertArrayClose(firstRow, singleResult, 1e-10, 'First row mismatch');
});

test('JMA batch - with NaN data', () => {

    const data = new Float64Array([NaN, NaN, ...Array.from({length: 20}, (_, i) => i + 1)]);

    const result = wasm.jma_batch(data, {
        period_range: [3, 5, 2],
        phase_range: [50.0, 50.0, 0],
        power_range: [2, 2, 0]
    });

    assert.strictEqual(result.rows, 2);
    assert.strictEqual(result.cols, data.length);


    for (let row = 0; row < 2; row++) {
        const rowStart = row * data.length;
        assert(isNaN(result.values[rowStart]), `Row ${row} index 0 should be NaN`);
        assert(isNaN(result.values[rowStart + 1]), `Row ${row} index 1 should be NaN`);

        assert(isFinite(result.values[rowStart + 2]), `Row ${row} index 2 should be finite`);
    }
});
